src/Discord/Internal/Types/Interactions.hs

```diff
@@ -0,0 +1,363 @@
+              ("value", toJSON <$> applicationCommandInteractionDataOptionValue),
+              ("options", toJSON <$> applicationCommandInteractionDataOptionOptions),
+              ("focused", toJSON <$> applicationCommandInteractionDataOptionFocused)
+            ]
+      ]
+
+instance FromJSON ApplicationCommandInteractionDataOption where
+  parseJSON =
+    withObject
+      "ApplicationCommandInteractionDataOption"
+      ( \v ->
+          ApplicationCommandInteractionDataOption
+            <$> v .: "name"
+            <*> v .: "type"
+            <*> v .:? "value"
+            <*> v .:? "options"
+            <*> v .:? "focused"
+      )
+
+-- | The data to respond to an interaction with. Unless specified otherwise, you
+-- only have three seconds to reply to an interaction before a failure state is
+-- given.
+data InteractionResponse = InteractionResponse
+  { interactionResponseType :: InteractionCallbackType,
+    interactionResponseData :: Maybe InteractionCallbackData
+  }
+  deriving (Show, Read, Eq)
+
+instance Default InteractionResponse where
+  def = InteractionResponse CHANNEL_MESSAGE_WITH_SOURCE Nothing
+
+instance ToJSON InteractionResponse where
+  toJSON InteractionResponse {..} =
+    object
+      [ (name, value)
+        | (name, Just value) <-
+            [ ("type", toMaybeJSON interactionResponseType),
+              ("data", toJSON <$> interactionResponseData)
+            ]
+      ]
+
+-- | What's the type of the response?
+data InteractionCallbackType
+  = -- | Responds to a PING.
+    PONG
+  | -- | Respond with a message to the interaction
+    CHANNEL_MESSAGE_WITH_SOURCE
+  | -- | Respond with a message to the interaction, after a delay. Sending this
+    -- back means the interaction token lasts for 15 minutes.
+    DEFERRED_CHANNEL_MESSAGE_WITH_SOURCE
+  | -- | For components, edit the original message later.
+    DEFERRED_UPDATE_MESSAGE
+  | -- | For components, edit the original message.
+    UPDATE_MESSAGE
+  | -- | Respond to an autocomplete interaction with suggested choices.
+    APPLICATION_COMMAND_AUTOCOMPLETE_RESULT
+  deriving (Show, Read, Eq, Data)
+
+instance Enum InteractionCallbackType where
+  fromEnum PONG = 1
+  fromEnum CHANNEL_MESSAGE_WITH_SOURCE = 4
+  fromEnum DEFERRED_CHANNEL_MESSAGE_WITH_SOURCE = 5
+  fromEnum DEFERRED_UPDATE_MESSAGE = 6
+  fromEnum UPDATE_MESSAGE = 6
+  fromEnum APPLICATION_COMMAND_AUTOCOMPLETE_RESULT = 6
+  toEnum a = fromJust $ lookup a table
+    where
+      table = makeTable PONG
+
+instance ToJSON InteractionCallbackType where
+  toJSON = toJSON . fromEnum
+
+-- | Convenience wrapper for two separate types of callback.
+data InteractionCallbackData = ICDM InteractionCallbackMessages | ICDA InteractionCallbackAutocomplete
+  deriving (Show, Read, Eq)
+
+instance ToJSON InteractionCallbackData where
+  toJSON (ICDM icdm) = toJSON icdm
+  toJSON (ICDA icda) = toJSON icda
+
+type InteractionCallbackAutocomplete = [ApplicationCommandOptionChoice]
```

Comment:
What's the use for this?

Reply:
It makes it clearer what the option is in terms of types; it can be removed if wanted.

Reply:
Hmm. Okay to leave for now.